community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -19,112 +19,302 @@
+     * and any work unit in the pile throws an exception. Thus the current thread is not
+     * gauranteed to observe any exception its unit of work might throw, since the
+     * exception will be thrown in whichever thread that ends up actually performing the work.
      */
-    public void apply( W work )
+    public void apply( W work ) throws ExecutionException
     {
         // Schedule our work on the stack.
-        WorkUnit<Material,W> unit = new WorkUnit<>( work );
-        unit.next = stack.getAndSet( unit ); // benign race, see reverse()
+        WorkUnit<Material,W> unit = enqueueWork( work );
 
         // Try grabbing the lock to do all the work, until our work unit
         // has been completed.
-        boolean wasInterrupted = false;
         int tryCount = 0;
         do
         {
             tryCount++;
-            try
+            checkFailure( tryDoWork( unit, tryCount, true ) );
+        }
+        while ( !unit.isDone() );
+    }
+
+    /**
+     * Apply the given work to the material in a thread-safe way, possibly asynchronously
+     * if contention is observed with other threads, and possibly by combining it with other work.
+     * <p>
+     * The returned future can be cancelled while it is still enqueued to run, but cancellation
+     * may race with work combining and application, and have slightly weaker semantics than what
+     * the {@link Future} interface specifies. for instance, a unit of that is cancelled, and where
+     * the {@link Future#cancel(boolean)} cancel} method returns {@code true}, may still and up
+     * being applied.
+     * The reason for this departure from specification is implementation efficiency of the other
+     * {@code WorkSync} features, since the cancellation feature is likely rarely used.
+     * <p>
+     * The given unit of work may be done by this thread, or any other thread that is concurrently
+     * submitting work to the {@code WorkSync}. If this unit of work, or any of the other units of
+     * work, throws an exception, then the exception will surface in the thread that ends up doing
+     * the work. This may manifest itself as an {@link ExecutionException} thrown from one of the
+     * {@code get} methods of {@link Future}, or from the {@link #apply(Work)} method.
+     *
+     * @param work The work to be done.
+     * @return A {@link Future} representing the eventual completion of the work.
+     */
+    public Future<?> applyAsync( W work )
+    {
+        // Schedule our work on the stack.
+        WorkUnit<Material,W> unit = enqueueWork( work );
+
+        // Make an attempt at doing the work immediately.
+        Throwable failure = tryDoWork( unit, 0, false );
+        if ( failure != null )
+        {
+            return new FutureThrow( failure );
+        }
+
+        // Otherwise return a future where the 'get' methods will do the work,
+        // if it hasn't been done already.
+        return new Future<Object>()
```

Comment:
The return type of `applyAsync` does not need to change to `Future<Void>`, but it could be changed as well.